nucleus/metrics/geometry.py

```diff
@@ -0,0 +1,184 @@
+from typing import List, Tuple, Union
+
+import numpy as np
+
+TOLERANCE = 1e-8
+
+
+class GeometryPoint:
+    def __init__(self, xy: Union[Tuple[float, float], np.ndarray]):
+        self.xy = np.array(xy)
+        self.x = xy[0]
+        self.y = xy[1]
+
+    def __repr__(self) -> str:
+        return f"GeometryPoint(xy=({self.xy[0]}, {self.xy[1]})"
+
+    def __add__(self, p: "GeometryPoint") -> "GeometryPoint":
+        return GeometryPoint(self.xy + p.xy)
+
+    def __sub__(self, p: "GeometryPoint") -> "GeometryPoint":
+        return GeometryPoint(self.xy - p.xy)
+
+    def __rmul__(self, scalar: float) -> "GeometryPoint":
+        return GeometryPoint(self.xy * scalar)
+
+    def __mul__(self, scalar: float) -> "GeometryPoint":
+        return GeometryPoint(self.xy * scalar)
+
+    # Operator @
+    def __matmul__(self, p: "GeometryPoint") -> float:
+        return self.xy @ p.xy
+
+    def length(self) -> float:
+        return np.linalg.norm(self.xy)
+
+    def cmp(self, p: "GeometryPoint") -> float:
+        return np.abs(self.xy - p.xy).sum()
+
+
+class GeometryPolygon:
+    def __init__(self, points: List[GeometryPoint]):
+        self.points = points
+
+        points_x = np.array([point.x for point in points])
+        points_y = np.array([point.y for point in points])
+        self.signed_area = (
+            points_x @ np.roll(points_y, 1) - points_x @ np.roll(points_y, -1)
+        ) / 2
+        self.area = np.abs(self.signed_area)
+
+    def __len__(self):
+        return len(self.points)
+
+    def __getitem__(self, idx):
+        return self.points[idx]
+
+    def __repr__(self) -> str:
+        points = ", ".join([str(point) for point in self.points])
+        return f"GeometryPolygon({points})"
+
+
+Segment = Tuple[GeometryPoint, GeometryPoint]
+
+
+# alpha * a1 + (1 - alpha) * a2 = beta * b1 + (1 - beta) * b2
+def _segment_intersection(
```

Comment:
Did you test this out? Can you add a simple unit test for this logic?

Reply:
Added unit tests for all of the methods